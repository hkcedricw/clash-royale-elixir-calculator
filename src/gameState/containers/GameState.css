.container {
  text-align: left;
}

.container * {
  margin-top: 5px;
  margin-bottom: 5px
}

.button-container {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.button-container button {
  margin-right: 3px
}

.button {
  height: 30px
}