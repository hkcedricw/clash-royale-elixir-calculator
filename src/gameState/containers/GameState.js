import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { 
  gameStartAction,
  gameStopAction,
  doubleElixirAction,
  addElixirAction,
  deductElixirAction,
  addPumpAction,
  destroyPumpAction
} from '../actions/gameStateAction';
import {
  DOUBLE_ELIXIR_TIME,
  NORMAL_ELIXIR_ADD_TIME,
  DOUBLE_ELIXIR_ADD_TIME,
  PUMP_ELIXIR
} from '../constants';
import ElixirDeductButton from '../components/ElixirDeductButton/ElixirDeductButton';
import DestroyPumpButton from '../components/DestroyPumpButton/DestroyPumpButton';
import './GameState.css';

const mapStateToProps = state => ({
  ...state.gameState
});

const mapDispatchToProps = dispatch => ({
  gameStartAction: () => gameStartAction(dispatch),
  gameStopAction: () => gameStopAction(dispatch),
  doubleElixirAction: () => doubleElixirAction(dispatch),
  addElixirAction: () => addElixirAction(dispatch),
  deductElixirAction: (number) => deductElixirAction(dispatch, number),
  addPumpAction: () => addPumpAction(dispatch),
  destroyPumpAction: (pumpId) => destroyPumpAction(dispatch, pumpId)
});

export class GameState extends Component {
  startGame = (event) => {
    this.props.gameStartAction();
    if(!this.gameTimerId) {
      this.gameTimerId = setTimeout(
        this.props.doubleElixirAction,
        DOUBLE_ELIXIR_TIME
      );
    }
    if(!this.addElixirTimerId) {
      this.addElixir();
    }
  }

  stopGame = (event) => {
    this.props.gameStopAction();
    if(this.gameTimerId) {
      clearTimeout(this.gameTimerId);
      this.gameTimerId = false;
    }
    if(this.addElixirTimerId) {
      clearTimeout(this.addElixirTimerId);
      this.addElixirTimerId = false;
    }
  }

  addElixir = () => {
    const add = () => {
      this.props.addElixirAction();
      this.addElixir();
    }
    this.addElixirTimerId = !this.props.doubleElixirStarted 
    ? setTimeout(
      add,
      NORMAL_ELIXIR_ADD_TIME
    )
    : setTimeout(
      add,
      DOUBLE_ELIXIR_ADD_TIME
    )
  }

  deductElixir = (number) => {
    if(this.props.elixirs >= number && this.props.started) {
      this.props.deductElixirAction(number);
    }
  }

  addPump = () => {
    if(this.props.elixirs >= PUMP_ELIXIR && this.props.started) {
      this.props.addPumpAction();
    }
  }

  destroyPump = (pumpId) => {
    clearInterval(pumpId);
    this.props.destroyPumpAction(pumpId);
  }

  render() {
    const elixirDeductButtons = [];
    for(var i=1; i<=9; i++){
      elixirDeductButtons.push((
        <div key={`elixirDeduct-${i}`}><ElixirDeductButton clickAction={this.deductElixir} number={i} /></div>
      ))
    }
    const pumps = [];
    this.props.pumpIds.forEach((pumpId) => {
      pumps.push((
        <div key={`destroyPump-${pumpId}`}><DestroyPumpButton clickAction={this.destroyPump} pumpId={pumpId} /></div>
      ));
    })
    return (
      <div className='container'>
        <div className='button-container'>
          <div>
            <button className='button' onClick={this.startGame}>Start</button>
          </div>
          <div>
            <button className='button' onClick={this.stopGame}>Stop</button>
          </div>
        </div>
        <div>
          Elixirs: {this.props.elixirs}
        </div>
        <div>
          Deduct Elixir:
        </div>
        <div className='button-container'>
          {elixirDeductButtons}
        </div>
        <div>
          Pumps:
        </div>
        <div className='button-container'>
          {pumps}
        </div>
        <div className='button-container'>
          <div>
            <button className='button' onClick={this.addPump}>Add Pump</button>
          </div>
        </div>
      </div>
    );
  }
}

GameState.propTypes = {
  started: PropTypes.bool,
  elixirs: PropTypes.number,
  doubleElixirStarted: PropTypes.bool,
  pumps: PropTypes.array,
  gameStartAction: PropTypes.func,
  gameStopAction: PropTypes.func,
  doubleElixirAction: PropTypes.func,
  addElixirAction: PropTypes.func,
  deductElixirAction: PropTypes.func,
  addPumpAction: PropTypes.func,
  destroyPumpAction: PropTypes.func
}

export default connect(mapStateToProps, mapDispatchToProps)(GameState);