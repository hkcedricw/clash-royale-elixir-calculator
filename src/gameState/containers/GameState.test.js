import React from 'react';
import { shallow } from 'enzyme';
import { GameState } from './GameState.js';

describe('GameState', () => {
  let defaultProps = {
    started: false,
    elixirs: 0,
    doubleElixirStarted: false,
    pumpIds: []
  }
  it('renders correctly', () => {
    const wrapper = shallow(<GameState {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});