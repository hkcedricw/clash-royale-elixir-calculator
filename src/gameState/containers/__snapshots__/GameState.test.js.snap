// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameState renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="button-container"
  >
    <div>
      <button
        className="button"
        onClick={[Function]}
      >
        Start
      </button>
    </div>
    <div>
      <button
        className="button"
        onClick={[Function]}
      >
        Stop
      </button>
    </div>
  </div>
  <div>
    Elixirs: 
    0
  </div>
  <div>
    Deduct Elixir:
  </div>
  <div
    className="button-container"
  >
    <div
      key="elixirDeduct-1"
    >
      <ElixirDeductButton
        clickAction={[Function]}
        number={1}
      />
    </div>
    <div
      key="elixirDeduct-2"
    >
      <ElixirDeductButton
        clickAction={[Function]}
        number={2}
      />
    </div>
    <div
      key="elixirDeduct-3"
    >
      <ElixirDeductButton
        clickAction={[Function]}
        number={3}
      />
    </div>
    <div
      key="elixirDeduct-4"
    >
      <ElixirDeductButton
        clickAction={[Function]}
        number={4}
      />
    </div>
    <div
      key="elixirDeduct-5"
    >
      <ElixirDeductButton
        clickAction={[Function]}
        number={5}
      />
    </div>
    <div
      key="elixirDeduct-6"
    >
      <ElixirDeductButton
        clickAction={[Function]}
        number={6}
      />
    </div>
    <div
      key="elixirDeduct-7"
    >
      <ElixirDeductButton
        clickAction={[Function]}
        number={7}
      />
    </div>
    <div
      key="elixirDeduct-8"
    >
      <ElixirDeductButton
        clickAction={[Function]}
        number={8}
      />
    </div>
    <div
      key="elixirDeduct-9"
    >
      <ElixirDeductButton
        clickAction={[Function]}
        number={9}
      />
    </div>
  </div>
  <div>
    Pumps:
  </div>
  <div
    className="button-container"
  />
  <div
    className="button-container"
  >
    <div>
      <button
        className="button"
        onClick={[Function]}
      >
        Add Pump
      </button>
    </div>
  </div>
</div>
`;
